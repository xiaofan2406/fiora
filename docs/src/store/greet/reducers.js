import { combineReducers } from 'redux';

import actionTypes from './actions';

export const initialMessage = 'World';

export function messageReducer(state = initialMessage, action) {
  switch (action.type) {
    case actionTypes.SET_MESSAGE:
      return action.payload;
    case actionTypes.RESET:
      return initialMessage;
    default:
      return state;
  }
}

export const initialTimes = 1;

export function timesReducer(state = initialTimes, action) {
  switch (action.type) {
    case actionTypes.SET_TIMES:
      return action.payload;
    case actionTypes.RESET:
      return initialTimes;
    default:
      return state;
  }
}

export default combineReducers({
  message: messageReducer,
  times: timesReducer
});
