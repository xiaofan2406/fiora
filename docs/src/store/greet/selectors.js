export const getMessage = state => state.greet.message;

export const getTimes = state => state.greet.times;

export const getGreeting = state =>
  Array(getTimes(state))
    .fill(getMessage(state))
    .join(' ');
