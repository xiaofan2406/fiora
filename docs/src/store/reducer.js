import { combineReducers } from 'redux';
import { reducer as fioraReducer } from 'fiora';

export default combineReducers({
  fiora: fioraReducer,
});
