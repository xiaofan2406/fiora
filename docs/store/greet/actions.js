const actionTypes = {
  SET_MESSAGE: 'greet/SET_MESSAGE',
  SET_TIMES: 'greet/SET_TIMES',
  RESET: 'greet/RESET',
};

export default actionTypes;

export const setMessage = payload => ({
  type: actionTypes.SET_MESSAGE,
  payload,
});

export const setTimes = payload => ({
  type: actionTypes.SET_TIMES,
  payload,
});

export const reset = () => ({
  type: actionTypes.RESET,
});
