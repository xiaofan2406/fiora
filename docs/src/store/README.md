# Redux Store

## Rules

Components should not be concerned with the the shape of the store, but they
should care about what _module_ and its actions and selectors that they need to
use.

Create an directory for each _module_ (separated depending on the business logic
of the application).

Each _module_ directory has at lease four files:

- **reducers.js** Reducer for this _module_, the reducer can be refactored into
  sub-reducer files.

- **actions.js** `export` the action _TYPES_ by default, and `export` each
  action separately.

- **selector.js** Each selector is aware of which key of the state it's living
  under. Therefore, there is no need for a global selector file.

- **index.js** `export` all actions, selectors and reducers. The default export
  is the reducer for this _module_. Each action or selector should be also
  exported (its very less likely they have the same name).

The root level `actions.js` and `selectors.js` exports all the functions created within all _modules_.

**actions** and **selectors** shared between _modules_ should use relative import.

## Usage

```js
import { setMessage, getMessage } from 'store/greet';

import { actions as greetActions } from 'store/greet';
import { selectors as greetSelectors } from 'store/greet';

import { actions } from 'store';
// actions.greet.setMessage
```
