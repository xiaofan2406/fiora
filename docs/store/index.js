import { createStore, applyMiddleware, compose } from 'redux';
import reduxThunk from 'redux-thunk';

import reducer from './reducer';
import { actions as greetActions, selectors as greetSelectors } from './greet';

export const actions = {
  greet: greetActions,
};

export const selectors = {
  greet: greetSelectors,
};

export const configureStore = (initialState = {}) => {
  const middlewares = [reduxThunk];

  if (process.env.NODE_ENV === 'development') {
    middlewares.push(require('redux-immutable-state-invariant').default());
    middlewares.push(require('redux-logger').default);
  }

  const store = createStore(
    reducer,
    initialState,
    compose(
      applyMiddleware(...middlewares),
      typeof window === 'object' &&
      typeof window.devToolsExtension !== 'undefined'
        ? window.devToolsExtension()
        : f => f
    )
  );

  if (module.hot) {
    module.hot.accept('./reducer', () => {
      const nextReducer = require('./reducer').default;
      store.replaceReducer(nextReducer);
    });
  }

  return store;
};

export default configureStore;
