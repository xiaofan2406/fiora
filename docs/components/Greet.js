import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { css } from 'react-emotion';
import { theme, fontSizes, spacing, verticalScroll } from 'styles';
import { actions, selectors } from 'store/greet';

const cssGreet = css`
  margin: 0px auto;
  width: 480px;
  & > .controls {
    text-align: center;
    & input {
      border: none;
      border-bottom: 1px solid ${theme.borderColor};
      outline: none;
      width: 100px;
      text-align: center;
      &:focus {
        border-bottom: 1px solid ${theme.primaryColor};
      }
    }
    & > button {
      margin-left: 1em;
      outline: none;
      background: none;
      border: 1px solid ${theme.borderColor};
      &:hover {
        border-color: ${theme.primaryColor};
        color: ${theme.primaryColor};
      }
    }
  }
  & > .result {
    height: 280px;
    ${verticalScroll};
    margin-top: ${spacing.external}px;
    font-size: ${fontSizes.large}px;
    font-family: cursive;
    &:before {
      content: '"';
    }
    &:after {
      content: '"';
    }
  }
`;

const Greet = ({ message, setMessage, times, setTimes, reset, greeting }) => {
  const handleTimesOnChange = event => {
    setTimes(+event.target.value || 0);
  };

  const handleMessageOnChange = event => {
    setMessage(event.target.value);
  };

  return (
    <div className={cssGreet}>
      <div className="controls">
        <span>Say </span>
        <input
          type="text"
          onChange={handleMessageOnChange}
          value={message}
        />{' '}
        <input
          type="number"
          min={1}
          onChange={handleTimesOnChange}
          value={times}
        />{' '}
        times
        <button onClick={reset} type="button">
          Clear
        </button>
      </div>
      <div className="result">Hello {greeting}</div>
    </div>
  );
};

Greet.propTypes = {
  message: PropTypes.string.isRequired,
  times: PropTypes.number.isRequired,
  greeting: PropTypes.string.isRequired,
  setMessage: PropTypes.func.isRequired,
  setTimes: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired,
};

export const mapStateToProps = state => ({
  message: selectors.getMessage(state),
  times: selectors.getTimes(state),
  greeting: selectors.getGreeting(state),
});

export { Greet as Component };

export default connect(
  mapStateToProps,
  {
    setMessage: actions.setMessage,
    setTimes: actions.setTimes,
    reset: actions.reset,
  }
)(Greet);
