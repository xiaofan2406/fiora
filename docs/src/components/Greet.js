import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { css } from 'emotion';
import { theme, colors, fontSizes, spacing } from 'styles';
import { actions, selectors } from 'store/greet';

const greet = css`
  margin: 0 auto;
  width: 460px;
  padding-top: ${spacing.externalBreak}px;
  & input {
    border: none;
    border-bottom: 1px solid ${colors.grey};
    outline: none;
    width: 100px;
    text-align: center;
    &:focus {
      border-bottom: 1px solid ${theme.primaryColor};
    }
  }
  & button {
    margin-left: 1em;
    outline: none;
    background: none;
    border: 1px solid ${colors.grey};
    &:hover {
      border-color: ${theme.primaryColor};
      color: ${theme.primaryColor};
    }
  }
  & .result {
    margin-top: ${spacing.externalBreak}px;
    font-size: ${fontSizes.large}px;
    font-family: cursive;
    &:before {
      content: '"';
    }
    &:after {
      content: '"';
    }
  }
`;

function Greet({ message, setMessage, times, setTimes, reset, greeting }) {
  const handleTimesOnChange = event => {
    setTimes(+event.target.value || 0);
  };

  const handleMessageOnChange = event => {
    setMessage(event.target.value);
  };

  return (
    <div className={greet}>
      <div>
        <span>Say </span>
        <input
          type="text"
          onChange={handleMessageOnChange}
          value={message}
        />{' '}
        <input
          type="number"
          min={1}
          onChange={handleTimesOnChange}
          value={times}
        />{' '}
        times
        <button onClick={reset}>Clear</button>
      </div>
      <div className="result">Hello {greeting}</div>
    </div>
  );
}

Greet.propTypes = {
  message: PropTypes.string.isRequired,
  times: PropTypes.number.isRequired,
  greeting: PropTypes.string.isRequired,
  setMessage: PropTypes.func.isRequired,
  setTimes: PropTypes.func.isRequired,
  reset: PropTypes.func.isRequired
};

export const mapStateToProps = state => ({
  message: selectors.getMessage(state),
  times: selectors.getTimes(state),
  greeting: selectors.getGreeting(state)
});

export { Greet as Component };

export default connect(mapStateToProps, {
  setMessage: actions.setMessage,
  setTimes: actions.setTimes,
  reset: actions.reset
})(Greet);
