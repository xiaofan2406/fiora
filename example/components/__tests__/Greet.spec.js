import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { selectors } from 'store/greet';
import { Component, mapStateToProps } from '../Greet';

const props = {
  message: 'World',
  setMessage: jest.fn(),
  times: 1,
  setTimes: jest.fn(),
  reset: jest.fn(),
  greeting: 'World'
};

test('snapshot is matching', () => {
  const component = shallow(<Component {...props} />);

  expect(toJson(component)).toMatchSnapshot();
});

test('mapStateToProps', () => {
  const mockState = { greet: {} };
  const getMessageSpy = jest.spyOn(selectors, 'getMessage');
  const getTimesSpy = jest.spyOn(selectors, 'getTimes');
  const getGreetingSpy = jest.spyOn(selectors, 'getGreeting');
  mapStateToProps(mockState);

  expect(getMessageSpy).toHaveBeenLastCalledWith(mockState);
  expect(getTimesSpy).toHaveBeenLastCalledWith(mockState);
  expect(getGreetingSpy).toHaveBeenLastCalledWith(mockState);
});

describe('prop: setMessage', () => {
  let component;
  let spy;

  beforeAll(() => {
    component = mount(<Component {...props} />);
    spy = jest.spyOn(props, 'setMessage');

    expect(spy).toHaveBeenCalledTimes(0);
  });

  test('called on the text input change event', () => {
    component
      .find('input[type="text"]')
      .simulate('change', { target: { value: '2' } });
    expect(spy).toHaveBeenCalledTimes(1);
  });

  test('called with event value', () => {
    component
      .find('input[type="text"]')
      .simulate('change', { target: { value: 'Hey' } });

    expect(spy).toHaveBeenLastCalledWith('Hey');
  });
});

describe('prop: setTimes', () => {
  let component;
  let spy;

  beforeAll(() => {
    component = mount(<Component {...props} />);
    spy = jest.spyOn(props, 'setTimes');

    expect(spy).toHaveBeenCalledTimes(0);
  });

  test('called on the number input change event', () => {
    component
      .find('input[type="number"]')
      .simulate('change', { target: { value: '2' } });

    expect(spy).toHaveBeenCalledTimes(1);
  });

  test('called with event value parsed as a number', () => {
    component
      .find('input[type="number"]')
      .simulate('change', { target: { value: '2' } });

    expect(spy).toHaveBeenLastCalledWith(2);
  });

  test('called with 0 when event value cannot be parsed to a number', () => {
    component.find('input[type="number"]').simulate('change', { target: {} });

    expect(spy).toHaveBeenLastCalledWith(0);
  });
});
