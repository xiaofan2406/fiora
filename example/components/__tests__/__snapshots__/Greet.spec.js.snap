// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot is matching 1`] = `
<div
  className="css-1yriv2r"
>
  <div>
    <span>
      Say 
    </span>
    <input
      onChange={[Function]}
      type="text"
      value="World"
    />
     
    <input
      min={1}
      onChange={[Function]}
      type="number"
      value={1}
    />
     
    times
    <button
      onClick={[Function]}
    >
      Clear
    </button>
  </div>
  <div
    className="result"
  >
    Hello 
    World
  </div>
</div>
`;
